### PR TITLE
Prepare docs and Swagger definition for automatically generated API docs

### PR DESCRIPTION
**Pull request by bfirsh**
**- What I did**

This is some preparation for automatically generating API documentation from a Swagger definition. See https://github.com/docker/docker/issues/22931.

It is the Engine half of [this documentation pull request](https://github.com/docker/docker.github.io/pull/606). This pull request should be merged before the documentation pull request.

It consists of a bunch of things:

- Removed the API reference for the latest API version. This is now all in `api/swagger.yaml`, which the documentation is automatically generated from.
- Removed all the other API documentation from this repository. The old versions barely ever change, and all the other stuff is not tied to any code change, so they should probably just live in docker/docker.github.io.
- Renamed "Remote API" to "Engine API". See https://github.com/docker/docker/issues/28319 for details.
- Update `api/swagger.yaml` with some more documentation and some recent API changes.

**- A picture of a cute animal (not mandatory but encouraged)**

robot-cat

/cc @dnephin @johndmulhausen

### CODE REVIEW

**Comment by johndmulhausen** [comment]
So excited about this, thanks @bfirsh. Slowly, we are killing all these hand-rolled markdown files and I couldn't be happier. To be getting such nice-looking API docs is just an added bonus!

cc @mstanleyjones

**Comment by thaJeztah** [comment]
Is there an easy way to generate these docs for preview / reviewing pull requests?

**Comment by bfirsh** [comment]
@thaJeztah I think the easiest way is to copy `swagger.yaml` into the right place in the documentation repo and run the documentation locally.

That could be documented somewhere I suppose. Perhaps a "API documentation" section in the contributing docs?

File: api/swagger.yaml [hunk 1]
```diff
@@ -6468,6 +6555,31 @@ paths:
             format: "binary"
       tags:
         - "Plugins"
+  /plugins/{name}/push:
+    post:
+      summary: "Push a plugin"
+      operationId: "PluginsPush"
```
Comment (dnephin) [inline api/swagger.yaml]: Could we make this `PluginPush` ? I've been normalizing everything to be singular.

File: api/swagger.yaml [hunk 2]
```diff
@@ -7067,6 +7246,84 @@ paths:
...
       tags:
         - "Services"
+  /services/{id}/logs:
+    get:
+      summary: "Get service logs"
+      description: |
+        Get `stdout` and `stderr` logs from a service.
+
+        **Note**: This endpoint works only for services with the `json-file` or `journald` logging drivers.
+      operationId: "ServicesLogs"
```
Comment (dnephin) [inline api/swagger.yaml]: `ServiceLogs`

File: api/swagger.yaml [hunk 3]
```diff
@@ -7238,3 +7495,149 @@ paths:
           type: "string"
       tags:
         - "Tasks"
+  /secrets:
+    get:
+      summary: "List secrets"
+      operationId: "SecretsList"
```
Comment (dnephin) [inline api/swagger.yaml]: `SecretList`

File: api/swagger.yaml [hunk 4]
```diff
@@ -7238,3 +7495,149 @@ paths:
...
+          description: |
+            A JSON encoded value of the filters (a `map[string][]string`) to process on the secrets list. Available filters:
+
+            - `names=<secret name>`
+      tags:
+        - "Secrets"
+  /secrets/create:
+    post:
+      summary: "Create a secret"
+      operationId: "SecretsCreate"
```
Comment (dnephin) [inline api/swagger.yaml]: `SecretCreate`

**Review by dnephin — COMMENTED** [review]
Some minor requests for changes to OperationID.

Thanks for adding all the missing endpoints.

LGTM

**Comment by bfirsh** [comment]
If we want this included in 1.13, then these changes need cherry-picking to the 1.13 branch. I see no reason why not to include it – it's purely a documentation change.

Following an offline discussion with @thaJeztah, I'm also going to add some docs about how to write and test the API docs.

**Comment by thaJeztah** [comment]
Adding to the milestone, but /cc @vieux

**Comment by bfirsh** [comment]
@thaJeztah Documentation added! https://github.com/bfirsh/docker/blob/prepare-api-docs-for-swagger-docs/project/API.md

**Review by mdlinville — APPROVED** [review]
This is so cool!! Thank you so much, @bfirsh !

**Review by thaJeztah — APPROVED** [review]
LGTM

File: api/README.md [hunk 5]
```diff
@@ -1,5 +1,42 @@
...
+
+`swagger.yaml` is validated by `hack/validate/swagger` to ensure it is a valid Swagger definition. This is useful for when you are making edits to ensure you are doing the right thing.
+
+## Viewing the API documentation
+
+When you make edits to `swagger.yaml`, you may want to check the generated API documentation to ensure it renders correctly.
+
+All the documentation generation is done in the documentation repository, [docker/docker.github.io](https://github.com/docker/docker.github.io). The Swagger definition is vendored periodically into this repository, but you can manually copy over the Swagger definition to test changes.
+
+Copy `api/swagger.yaml` in this repository to `engine/api/v1.25/swagger.yaml` in the documentation repository, overwriting what is already there. Then, run `docker-compose up` in the documentation repository and browse to [http://localhost:4000/engine/api/v1.25/](http://localhost:4000/engine/api/v1.25/) when it finishes rendering.
```
Comment (mlaventure) [inline api/README.md]: Replace `v1.25` with something like `<API_VERSION>` so we don't have to update this file every release? (we've already updated master to `v1.26`)

**Review by mlaventure — CHANGES_REQUESTED** [review]

File: api/README.md (hunk 5 above)
Comment (bfirsh) [inline api/README.md]: Heh, good point. I'll just remove the version number actually, because `/engine/api/` redirects to `/engine/api/v1.25/`!

Comment (bfirsh) [inline api/README.md]: @mlaventure Done!

**Review by mlaventure — APPROVED** [review]

**Comment by vieux** [comment]
LGTM

**Comment by thaJeztah** [comment]
There we go!

**Comment by vieux** [comment]
@bfirsh can you create a PR against the 1.13.x branch directly ? this one is tricky to cherry-pick since it's 5 commits and the version changed in master

**Comment by vieux** [comment]
@bfirsh nevermind I got it

**Comment by bfirsh** [comment]
@vieux Nice, thank you! Appreciate that, because I realise it was non-trivial...

**Comment by vieux** [comment]
@bfirsh https://github.com/docker/docker/blob/master/docs/api/version-history.md has 2 broken links, 1.25 and 1.26

**Comment by bfirsh** [comment]
@vieux They will exist once https://github.com/docker/docker.github.io/pull/606 exists. :)

Well, 1.26 won't, but the `1.13` branch doesn't have that broken link.

**Pull request merged**